src/utils/observability/src/axum.rs

```diff
@@ -0,0 +1,157 @@
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use http::Uri;
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+pub fn http_layer() -> tower_http::trace::TraceLayer<
+    tower_http::classify::SharedClassifier<tower_http::classify::ServerErrorsAsFailures>,
+    MakeSpan,
+    OnRequest,
+    OnResponse,
+> {
+    tower_http::trace::TraceLayer::new_for_http()
+        .on_request(OnRequest)
+        .on_response(OnResponse)
+        .make_span_with(MakeSpan)
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Clone, Debug)]
+pub struct OnRequest;
+
+impl<B> tower_http::trace::OnRequest<B> for OnRequest {
+    fn on_request(&mut self, request: &http::Request<B>, _: &tracing::Span) {
+        tracing::info!(
+            uri = %request.uri(),
+            version = ?request.version(),
+            headers = ?request.headers(),
+            "HTTP request",
+        );
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Clone, Debug)]
+pub struct OnResponse;
+
+impl<B> tower_http::trace::OnResponse<B> for OnResponse {
+    fn on_response(
+        self,
+        response: &http::Response<B>,
+        latency: std::time::Duration,
+        _span: &tracing::Span,
+    ) {
+        tracing::info!(
+            status = response.status().as_u16(),
+            headers = ?response.headers(),
+            latency = %Latency(latency),
+            "HTTP response"
+        );
+    }
+}
+
+struct Latency(std::time::Duration);
+
+impl std::fmt::Display for Latency {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{} ms", self.0.as_millis())
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Debug, Clone)]
+pub struct MakeSpan;
+
+impl<B> tower_http::trace::MakeSpan<B> for MakeSpan {
+    fn make_span(&mut self, request: &http::Request<B>) -> tracing::Span {
+        use opentelemetry::trace::TraceContextExt as _;
+        use tracing_opentelemetry::OpenTelemetrySpanExt as _;
+
+        // TODO: Extract parent context from the request
+
+        let method = request.method();
+        let route = RouteOrUri::from(request);
+
+        let span = tracing::info_span!(
+            "http_request",
+            %method,
+            %route,
+            // Placeholders for OTEL fileds that will be populated after span creation
+            trace_id = tracing::field::Empty,
+            "otel.name" = tracing::field::Empty,
+        );
+
+        // Extract trace ID from the OTEL context and add it to the tracing span
+        let context = span.context();
```

Comment:
Note for future readers:

https://docs.rs/tracing-opentelemetry/latest/tracing_opentelemetry/trait.OpenTelemetrySpanExt.html#tymethod.context